File: packages/tracelet_platform_interface/test/algorithms/battery_budget_engine_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:tracelet_platform_interface/tracelet_platform_interface.dart';


void main() async {
  await RustLib.init();
  group('BatteryBudgetEngine', () {
    test('constructor sets defaults', () {
      final engine = BatteryBudgetEngine();

      expect(engine.targetBudgetPerHour, 3.0);
      expect(engine.distanceFilter, 10.0);
      expect(engine.accuracyIndex, 0);
      expect(engine.periodicInterval, isNull);
    });

    test('constructor accepts custom initial values', () {
      final engine = BatteryBudgetEngine(
        targetBudgetPerHour: 5,
        initialDistanceFilter: 50,
        initialAccuracyIndex: 2,
        initialPeriodicInterval: 300,
      );

      expect(engine.targetBudgetPerHour, 5.0);
      expect(engine.distanceFilter, 50.0);
      expect(engine.accuracyIndex, 2);
      expect(engine.periodicInterval, 300);
    });

    test('clamps accuracy index to 0-4 range', () {
      final tooHigh = BatteryBudgetEngine(initialAccuracyIndex: 10);
      expect(tooHigh.accuracyIndex, 4);

      final tooLow = BatteryBudgetEngine(initialAccuracyIndex: -5);
      expect(tooLow.accuracyIndex, 0);
    });

    test('first processSample returns null (baseline)', () {
      final engine = BatteryBudgetEngine();

      final result = engine.processSample(0.95);
      expect(result, isNull);
    });

    test('second processSample within 60s returns null', () {
      final engine = BatteryBudgetEngine();

      // First call sets baseline
      engine.processSample(0.95);

      // Immediate second call — too soon (< 60 seconds)
      final result = engine.processSample(0.90);
      expect(result, isNull);
    });

    test('reset allows re-establishing baseline', () {
      final engine = BatteryBudgetEngine();

      engine.processSample(0.95);
      engine.reset();

      // After reset, next call should return null (new baseline)
      final result = engine.processSample(0.90);
      expect(result, isNull);
    });

    test('reset does not change configuration', () {
      final engine = BatteryBudgetEngine(
        targetBudgetPerHour: 5,
        initialDistanceFilter: 50,
        initialAccuracyIndex: 2,
      );

      engine.processSample(0.95);
      engine.reset();

      expect(engine.targetBudgetPerHour, 5.0);
      expect(engine.distanceFilter, 50.0);
      expect(engine.accuracyIndex, 2);
    });
  });

  group('BudgetAdjustmentEvent', () {
    test('constructor stores all fields', () {
      const event = BudgetAdjustmentEvent(
        currentBatteryDrain: 4.5,
        targetBudget: 3,
        newDistanceFilter: 50,
        newDesiredAccuracy: 2,
        newPeriodicInterval: 600,
      );

      expect(event.currentBatteryDrain, 4.5);
      expect(event.targetBudget, 3.0);
      expect(event.newDistanceFilter, 50.0);
      expect(event.newDesiredAccuracy, 2);
      expect(event.newPeriodicInterval, 600);
    });

    test('newPeriodicInterval defaults to null', () {
      const event = BudgetAdjustmentEvent(
        currentBatteryDrain: 2,
        targetBudget: 3,
        newDistanceFilter: 10,
        newDesiredAccuracy: 0,
      );

      expect(event.newPeriodicInterval, isNull);
    });
  });

  group('BatteryBudgetEngine — internal state', () {
    test('initial distanceFilter matches constructor param', () {
      final engine = BatteryBudgetEngine(initialDistanceFilter: 100);
      expect(engine.distanceFilter, 100.0);
    });

    test('initial periodicInterval matches constructor param', () {
      final engine = BatteryBudgetEngine(initialPeriodicInterval: 300);
      expect(engine.periodicInterval, 300);
    });

    test('periodicInterval is null when not provided', () {
      final engine = BatteryBudgetEngine();
      expect(engine.periodicInterval, isNull);
    });

    test('accuracy index clamped to minimum 0', () {
      final engine = BatteryBudgetEngine(initialAccuracyIndex: -100);
      expect(engine.accuracyIndex, 0);
    });

    test('accuracy index clamped to maximum 4', () {
      final engine = BatteryBudgetEngine(initialAccuracyIndex: 100);
      expect(engine.accuracyIndex, 4);
    });

    test('reset preserves distance filter and accuracy', () {
      final engine = BatteryBudgetEngine(
        targetBudgetPerHour: 5,
        initialDistanceFilter: 200,
        initialAccuracyIndex: 3,
        initialPeriodicInterval: 600,
      );

      // Establish baseline then reset
      engine.processSample(0.95);
      engine.reset();

      // Configuration should be preserved
      expect(engine.distanceFilter, 200.0);
      expect(engine.accuracyIndex, 3);
      expect(engine.periodicInterval, 600);

      // Next processSample should return null (new baseline)
      expect(engine.processSample(0.90), isNull);
    });

    test('charging (battery increase) returns null', () {
      final engine = BatteryBudgetEngine();

      // The engine uses DateTime.now() internally so we can only test
      // the first-sample baseline behavior directly. Charging detection
      // requires elapsed time > 60s which can't be simulated without
      // injecting a clock.
      engine.processSample(0.50);
      // Immediately calling with higher level returns null (too soon)
      expect(engine.processSample(0.60), isNull);
    });
  });

  group('BatteryBudgetEngine — time-dependent (clock injection)', () {
    late DateTime fakeNow;
    late BatteryBudgetEngine engine;

    setUp(() {
      fakeNow = DateTime(2024, 1, 1, 12);
    });

    BatteryBudgetEngine createEngine({
      double targetBudgetPerHour = 3.0,
      double initialDistanceFilter = 10.0,
      int initialAccuracyIndex = 0,
      int? initialPeriodicInterval,
    }) {
      return BatteryBudgetEngine(
        targetBudgetPerHour: targetBudgetPerHour,
        initialDistanceFilter: initialDistanceFilter,
        initialAccuracyIndex: initialAccuracyIndex,
        initialPeriodicInterval: initialPeriodicInterval,
        clock: () => fakeNow,
      );
    }

    test('throttles when draining too fast', () {
      engine = createEngine();

      // Baseline at 95%
      engine.processSample(0.95);

      // Advance 1 hour
      fakeNow = fakeNow.add(const Duration(hours: 1));

      // Battery dropped to 85% → 10% drain/hr >> 3% budget
      final result = engine.processSample(0.85);

      expect(result, isNotNull);
      expect(result!.currentBatteryDrain, closeTo(10.0, 0.1));
      expect(result.targetBudget, 3.0);
      // Distance filter should increase (10 * 1.5 = 15)
      expect(engine.distanceFilter, closeTo(15.0, 0.1));
      // Accuracy should degrade (0 → 1)
      expect(engine.accuracyIndex, 1);
    });

    test('boosts when under budget', () {
      engine = createEngine(
        targetBudgetPerHour: 10,
        initialDistanceFilter: 100,
        initialAccuracyIndex: 3,
      );

      // Baseline at 95%
      engine.processSample(0.95);

      // Advance 1 hour
      fakeNow = fakeNow.add(const Duration(hours: 1));

      // Battery dropped to 94% → 1% drain/hr << 10% budget
      final result = engine.processSample(0.94);

      expect(result, isNotNull);
      expect(result!.currentBatteryDrain, closeTo(1.0, 0.1));
      // Distance filter should decrease (100 * 0.8 = 80)
      expect(engine.distanceFilter, closeTo(80.0, 0.1));
      // Accuracy should improve (3 → 2)
      expect(engine.accuracyIndex, 2);
    });

    test('no adjustment when within error threshold', () {
      engine = createEngine();

      // Baseline at 95%
      engine.processSample(0.95);

      // Advance 1 hour
      fakeNow = fakeNow.add(const Duration(hours: 1));

      // Battery dropped to 92% → 3% drain/hr == exactly on budget
      final result = engine.processSample(0.92);

      // Error = 3.0 - 3.0 = 0.0 < 0.5 threshold → no adjustment
      expect(result, isNull);
    });

    test('no adjustment when charging', () {
      engine = createEngine();

      engine.processSample(0.50);

      fakeNow = fakeNow.add(const Duration(hours: 1));

      // Battery went UP (charging)
      final result = engine.processSample(0.60);
      expect(result, isNull);
    });

    test('too-soon sample is ignored', () {
      engine = createEngine();

      engine.processSample(0.95);

      // Only 30 seconds later — too soon
      fakeNow = fakeNow.add(const Duration(seconds: 30));

      final result = engine.processSample(0.50);
      expect(result, isNull);
    });

    test('throttles periodic interval when draining fast', () {
      engine = createEngine(initialPeriodicInterval: 300);

      engine.processSample(0.95);
      fakeNow = fakeNow.add(const Duration(hours: 1));

      // 10% drain/hr >> 3% budget
      final result = engine.processSample(0.85);

      expect(result, isNotNull);
      // 300 * 1.5 = 450
      expect(engine.periodicInterval, 450);
    });

    test('boosts periodic interval when under budget', () {
      engine = createEngine(
        targetBudgetPerHour: 10,
        initialPeriodicInterval: 300,
      );

      engine.processSample(0.95);
      fakeNow = fakeNow.add(const Duration(hours: 1));

      // 1% drain/hr << 10% budget
      final result = engine.processSample(0.94);

      expect(result, isNotNull);
      // 300 * 0.8 = 240
      expect(engine.periodicInterval, 240);
    });

    test('accuracy index does not exceed 4 on repeated throttling', () {
      engine = createEngine(targetBudgetPerHour: 1, initialAccuracyIndex: 3);

      engine.processSample(0.95);
      fakeNow = fakeNow.add(const Duration(hours: 1));
      engine.processSample(0.85); // accuracy 3 → 4

      expect(engine.accuracyIndex, 4);

      // Throttle again — accuracy already at max
      fakeNow = fakeNow.add(const Duration(hours: 1));
      engine.processSample(0.75);
      expect(engine.accuracyIndex, 4);
    });

    test('accuracy index does not go below 0 on repeated boosting', () {
      engine = createEngine(targetBudgetPerHour: 50, initialAccuracyIndex: 1);

      engine.processSample(0.95);
      fakeNow = fakeNow.add(const Duration(hours: 1));
      engine.processSample(0.94); // accuracy 1 → 0

      expect(engine.accuracyIndex, 0);

      fakeNow = fakeNow.add(const Duration(hours: 1));
      engine.processSample(0.93);
      expect(engine.accuracyIndex, 0);
    });

    test('distance filter clamped to min on aggressive boosting', () {
      engine = createEngine(targetBudgetPerHour: 50, initialDistanceFilter: 11);

      engine.processSample(0.95);
      fakeNow = fakeNow.add(const Duration(hours: 1));
      engine.processSample(0.94);

      // 11 * 0.8 = 8.8 → clamped to 10.0
      expect(engine.distanceFilter, 10.0);
    });

    test('distance filter clamped to max on aggressive throttling', () {
      engine = createEngine(
        targetBudgetPerHour: 0.1,
        initialDistanceFilter: 4000,
      );

      engine.processSample(0.95);
      fakeNow = fakeNow.add(const Duration(hours: 1));
      engine.processSample(0.85);

      // 4000 * 1.5 = 6000 → clamped to 5000
      expect(engine.distanceFilter, 5000.0);
    });

    test('periodic interval clamped to min 60 on boosting', () {
      engine = createEngine(
        targetBudgetPerHour: 50,
        initialPeriodicInterval: 65,
      );

      engine.processSample(0.95);
      fakeNow = fakeNow.add(const Duration(hours: 1));
      engine.processSample(0.94);

      // 65 * 0.8 = 52 → clamped to 60
      expect(engine.periodicInterval, 60);
    });

    test('periodic interval clamped to max 43200 on throttling', () {
      engine = createEngine(
        targetBudgetPerHour: 0.1,
        initialPeriodicInterval: 40000,
      );

      engine.processSample(0.95);
      fakeNow = fakeNow.add(const Duration(hours: 1));
      engine.processSample(0.85);

      // 40000 * 1.5 = 60000 → clamped to 43200
      expect(engine.periodicInterval, 43200);
    });

    test('multiple successive adjustments accumulate', () {
      engine = createEngine(targetBudgetPerHour: 1);

      // Round 1: heavy drain
      engine.processSample(0.95);
      fakeNow = fakeNow.add(const Duration(hours: 1));
      engine.processSample(0.85);

      expect(engine.distanceFilter, closeTo(15.0, 0.1)); // 10 * 1.5
      expect(engine.accuracyIndex, 1);

      // Round 2: still draining heavy
      fakeNow = fakeNow.add(const Duration(hours: 1));
      engine.processSample(0.75);

      expect(engine.distanceFilter, closeTo(22.5, 0.1)); // 15 * 1.5
      expect(engine.accuracyIndex, 2);
    });
  });
}

File: sdk/android/tracelet-sdk/src/test/kotlin/com/ikolvi/tracelet/sdk/algorithm/BatteryBudgetEngineTest.kt
package com.ikolvi.tracelet.sdk.algorithm

import org.junit.Before
import org.junit.Test
import kotlin.test.assertEquals
import kotlin.test.assertNotNull
import kotlin.test.assertNull
import kotlin.test.assertTrue

/**
 * Unit tests for [BatteryBudgetEngine].
 */
class BatteryBudgetEngineTest {

    private lateinit var engine: BatteryBudgetEngine

    @Before
    fun setUp() {
        engine = BatteryBudgetEngine(
            targetBudgetPerHour = 5.0,
            initialDistanceFilter = 50.0,
            initialAccuracyIndex = 0,
        )
    }

    @Test
    fun `first sample returns null (baseline)`() {
        val result = engine.processSample(0.85)
        assertNull(result, "First sample should be baseline, no adjustment")
    }

    @Test
    fun `sample too soon returns null`() {
        engine.processSample(0.85)
        // Immediately call again — less than 60s elapsed
        val result = engine.processSample(0.84)
        assertNull(result)
    }

    @Test
    fun `no adjustment when within error threshold`() {
        // 5%/hr target. Error threshold is 0.5%/hr.
        // 0.42% drop in 5 min = 5.04%/hr → error = 0.04%/hr < 0.5 threshold → no adjustment
        val result = simulateSamples(
            startLevel = 0.90,
            endLevel = 0.8958, // (0.90 - 0.8958) * 100 = 0.42%, * (3600/300) = 5.04%/hr
            elapsedMs = 5 * 60 * 1000L,
        )
        assertNull(result, "Should not adjust when drain is within threshold of target")
    }

    @Test
    fun `throttle when draining too fast`() {
        val result = simulateSamples(
            startLevel = 0.90,
            endLevel = 0.85, // 5% drop in 5 min = 60%/hr — way over 5% budget
            elapsedMs = 5 * 60 * 1000L,
        )
        assertNotNull(result)
        assertTrue(result.newDistanceFilter > 50.0, "Should increase distance filter: ${result.newDistanceFilter}")
        assertTrue(result.newDesiredAccuracy > 0, "Should degrade accuracy: ${result.newDesiredAccuracy}")
        assertTrue(result.currentBatteryDrain > engine.targetBudgetPerHour)
    }

    @Test
    fun `boost when under budget`() {
        val result = simulateSamples(
            startLevel = 0.90,
            endLevel = 0.899, // 0.1% in 5 min = 1.2%/hr — well under 5% budget
            elapsedMs = 5 * 60 * 1000L,
        )
        assertNotNull(result)
        assertTrue(result.newDistanceFilter < 50.0, "Should decrease distance filter: ${result.newDistanceFilter}")
    }

    @Test
    fun `charging returns null`() {
        val result = simulateSamples(
            startLevel = 0.80,
            endLevel = 0.85, // Battery went UP (charging)
            elapsedMs = 5 * 60 * 1000L,
        )
        assertNull(result, "Should not adjust while charging")
    }

    @Test
    fun `distance filter clamped to min`() {
        // Create engine with very small distanceFilter
        val smallEngine = BatteryBudgetEngine(
            targetBudgetPerHour = 50.0, // Very high budget
            initialDistanceFilter = 10.0,
            initialAccuracyIndex = 0,
        )
        val result = simulateSamples(
            engine = smallEngine,
            startLevel = 0.999,
            endLevel = 0.998, // Tiny drain → boost
            elapsedMs = 5 * 60 * 1000L,
        )
        if (result != null) {
            assertTrue(result.newDistanceFilter >= 10.0, "Distance filter should not go below 10m")
        }
    }

    @Test
    fun `distance filter clamped to max`() {
        val largeEngine = BatteryBudgetEngine(
            targetBudgetPerHour = 0.01, // Extremely tight budget
            initialDistanceFilter = 4000.0,
            initialAccuracyIndex = 4,
        )
        val result = simulateSamples(
            engine = largeEngine,
            startLevel = 0.90,
            endLevel = 0.80, // 10% drain in 5 min = 120%/hr — massive overshoot
            elapsedMs = 5 * 60 * 1000L,
        )
        if (result != null) {
            assertTrue(result.newDistanceFilter <= 5000.0, "Distance filter should not exceed 5000m")
        }
    }

    @Test
    fun `accuracy index clamped to 0-4`() {
        assertEquals(0, engine.accuracyIndex)
        // Throttle to increase accuracy index
        simulateSamples(
            startLevel = 0.90, endLevel = 0.80, elapsedMs = 5 * 60 * 1000L
        )
        assertTrue(engine.accuracyIndex in 0..4)
    }

    @Test
    fun `reset clears state`() {
        engine.processSample(0.90)
        engine.reset()
        // After reset, next sample should be baseline again
        val result = engine.processSample(0.89)
        assertNull(result, "After reset, first sample should be baseline")
    }

    @Test
    fun `periodic interval adjusted when present`() {
        val periodicEngine = BatteryBudgetEngine(
            targetBudgetPerHour = 5.0,
            initialDistanceFilter = 50.0,
            initialAccuracyIndex = 0,
            initialPeriodicInterval = 900, // 15 min
        )
        val result = simulateSamples(
            engine = periodicEngine,
            startLevel = 0.90,
            endLevel = 0.80, // Big drain → throttle
            elapsedMs = 5 * 60 * 1000L,
        )
        assertNotNull(result)
        assertNotNull(result.newPeriodicInterval)
        assertTrue(result.newPeriodicInterval!! > 900, "Should increase periodic interval")
    }

    @Test
    fun `periodic interval null when not periodic`() {
        val result = simulateSamples(
            startLevel = 0.90,
            endLevel = 0.80,
            elapsedMs = 5 * 60 * 1000L,
        )
        assertNotNull(result)
        assertNull(result.newPeriodicInterval, "Non-periodic engine should have null interval")
    }

    @Test
    fun `budget adjustment event contains correct target`() {
        val result = simulateSamples(
            startLevel = 0.90, endLevel = 0.80, elapsedMs = 5 * 60 * 1000L
        )
        assertNotNull(result)
        assertEquals(5.0, result.targetBudget)
    }

    // =========================================================================
    // Helpers
    // =========================================================================

    /**
     * Simulate two samples with controlled timing via reflection.
     */
    private fun simulateSamples(
        startLevel: Double,
        endLevel: Double,
        elapsedMs: Long,
        engine: BatteryBudgetEngine = this.engine,
    ): BudgetAdjustmentEvent? {
        // First sample (baseline)
        engine.processSample(startLevel)

        // Manipulate prevSampleTimeMs to simulate elapsed time
        val prevTimeField = BatteryBudgetEngine::class.java.getDeclaredField("prevSampleTimeMs")
        prevTimeField.isAccessible = true
        val now = System.currentTimeMillis()
        prevTimeField.set(engine, now - elapsedMs)

        return engine.processSample(endLevel)
    }
}
